### ISSUE TITLE
How to set up BBB to be compatible with auto-scaling env  

### ISSUE THREAD

**Issue by mabras** [bigbluebutton/bigbluebutton#8685]
I am planning to use the AWS auto-scaling infrastructure. I need your help to give an opinion about my approach:
BBB core server -> a decent server (c5 family instance)
recording -> s3
greenlight database (postgres) and mongoDB database -> separate server

What do you think about this? Is this the right service distributions?

Thanks

**Comment by ffdixon** [comment]
BigBlueButton is not designed for auto-scaling.  Unlike a stateless web application that uses an external database to keep its state, for example, BigBlueButton is a statefull application that has many local databases.

We (the core developers) haven't tried to do this yet.  Our focus has been to make the single BigBlueButton server as scalable as possible. 

How many concurrent users do you need to support at one time?

**Comment by mabras** [comment]
We do not know exactly. I am not concerned about the number of users per conference (the average is 100 users). My concern is how to serve, say, five conference with 100 concurrent users at the same time?

Thanks for your help.

**Comment by mabras** [comment]
Hello @capilkey ,
Can you give a guidance on this? Really appreciated.

**Comment by ffdixon** [comment]
@mabras We don't have a solution for auto-scaling, but we're thinking of how we might approach this differently.

We'll update this ticket when we have more to share.

**Comment by capilkey** [comment]
I don't really have anything much else to add that @ffdixon hasn't already said.

I think most of the companies that have set up BigBlueButton so that it can support larger numbers of meetings have a load balancer in front of multiple pre-configured BBB servers. The LMS sends the API calls to the load balancer and then the load balancer is responsible for assigning meetings to certain servers and serving as a proxy for further API calls.

**Comment by virtualrobotix** [comment]
So there are already some LMS that support this approach ? Like moodle x example ? 
If we use a different path x server could be possible generate a different rule respect of availability of resource on BBB server planned by a calendar x example ...  Do you think that could be a simple strategy to apply in real scenario ? 
Best
Roberto

**Comment by einhirn** [comment]
Look at https://github.com/blindsidenetworks/scalelite/ which can be integrated in LMS like a regular BBB Server (I don't know about LTI, though) - just change the API URL and Shared Secret to match the ScaleLite box.

**Comment by jon-hasen** [comment]
@ffdixon @capilkey  This is an issue we really would appreciate some focus on if you get a chance because who better to work on an auto-scaling solution than the people who developed BBB.

**Comment by GhaziTriki** [comment]
Hi @jon-hasen while auto-scaling is not part of BigBlueButton if you have cloud engineers you could achieve it. Any one who is used to large deployments can understand how BigBlueButton works from the official documentation and work on auto-scaling.

**Comment by jon-hasen** [comment]
@GhaziTriki thanks for the response, we have limited resources and so not as easy to do when we don't know the code ourselves.

**Comment by congthang1** [comment]
> This is what @capilkey said, basically, but with a link:
> Look at https://github.com/blindsidenetworks/scalelite/ which can be integrated in LMS like a regular BBB Server (I don't know about LTI, though) - just change the API URL and Shared Secret to match the ScaleLite box.
> You then register multiple BBB servers with Scalelite and it distributes the rooms/meetings/sessions across all of them.

If we can separate these modules like Scalelite doing:
1. BigBlueButton Servers
2. NFS Shared Volume
3. PostgreSQL Database
4. Redis Cache

I think we don't even need Scalelite. As we can use AWS ECS autoscaling, setting up share volume for all BigBlueButton server container (can use S3 even) then put them under a autoscaling group. 
All BigBlueButton Servers will have same secret. AWS autoscaling will add new BigBlueButton Server and route the traffic to each BigBlueButton Server automatically on demand. 
am I missing something?

**Comment by vlasov01** [comment]
I'm wondering if someone captured performance metrics for different components of BBB under the load?
I've captured some data from running BBB 2.2.11 on a box with 8 cores and 4GB of RAM using top utility (by CPU usage):
![image](https://user-images.githubusercontent.com/1688219/82965767-1cc87a00-9f97-11ea-941b-24157f274b1f.png)

and sorted by memory usage:
![image](https://user-images.githubusercontent.com/1688219/82965779-2356f180-9f97-11ea-9ed5-f6e46dc8c2cc.png)

Based on these numbers freeswitch is the busiest component by far. If it is so tightly integrated  with other components it can be scaled horizontally. There is some work in progress on making a Docker image for BBB. Once it is completed it should be a relatively straightforward task to scale it horizontally. 
I was thinking about moving the management components of BBB like Greenlight with PostgreSQL to another box so I can run multiple BBB with just one Greenlight . Does it make sense?

**Comment by einhirn** [comment]
That depends - when you have more video streams, Kurento takes over.

Anyway - yes, that is exactly what we're doing with Scalelite: I have one host where I installed Scalelite as Loadbalancer for BBB and it also hosts Greenlight and it's Postgres DB. See here for some Idea about it: https://github.com/blindsidenetworks/scalelite/issues/83

**Comment by abelperezok** [comment]
@einhirn @vlasov01 That's really interesting insights you're sharing here. I'm also looking into scaling horizontally and by the looks of it, Scalelite has done to a good level. 
However, in an ideal scenario as @congthang1 mentioned, it would be great if some of the components can be scaled independently either using containers or just servers in a way that the state is outside the server and therefore they become stateless allowing services like AWS auto scaling and ECS to create and destroy servers/containers as the demand goes up and down.
@einhirn could you share any resource you've come across about pushing many users in one box? 
I understand that usage determines that number and I'd like to get more insight on it. Seems like you've had some experience setting up Scalelite.

**Comment by einhirn** [comment]
@abelperezok Mainly I've just heard colleagues from other universities talking about it.
I just looked at the [Architecture description for BBB](https://docs.bigbluebutton.org/2.2/architecture.html)  and also at my servers BBB was installed on and found that there are some components that use a Node.JS server process - which only is single threaded as far as I know. I had the idea to run multiple instances of such a server processes on different ports and use Nginx to load balance between them - and it seems to work, at least in my test environment.

Our own lectures are almost always inside the limits of one BBB-Server, so I didn't need to optimize it and didn't gather any experience with it...

There's already a pull request to split up the BBB-HTML5-UI-Server into 3 Parts to distribute the load, which is a different approach from what I tried... #8788

Concerning FreeSWITCH's load: I don't know yet how that scales, but I guess to a certain degree it depends on the users connected "with microphone" as opposed to "listen only". I don't know whether there's a difference if they're all in one session (ie. FreeSwitch conference or voice bridge) or if the users are connected to a number of different conference bridges...

With Kurento it's quite simple: It has to process roughly <users with Webcams enabled> to the power of 2 streams for each session - so it will use a lot of CPU when there are many webcams enabled on a server. (see [WebRTCGlossary: SFU](https://webrtcglossary.com/sfu/))

I don't know about the performance capabilities of the java-based services. But I see two further places where Node.JS is used:
  * bbb-webrtc-sfu - I guess to manage Kurento
  * Etherpad-lite - for shared notes.

At some point it boils down to what amount of messages redis is able to pass between components and how many of such messages are generated. Or MongoDB, in case of HTML5-UI, which additionally uses that.
And then there's also the client's browser that has to process all information sent to it... For Webcams this oftentimes is the bottleneck...

**Comment by congthang1** [comment]
@einhirn yes I see these components can be  split and place somewhere out of sever but yes again question is how to autoscale each one. So like FreeSwitch if its statefull then problem not completely solved. Scalite can help us to autoscale by some more work on how to connect autoscaling cluster with Scalite rake cli. But its about autoscale a lot of small rooms (like 200 concurrences). Ideally is scalIing to a big room like 3000 and we just add new severs on demand.

**Comment by matiasilva** [comment]
At that point, there are just better tools for that purpose. No video conferencing tool should ever need to host 3000 users in one room. Why would you even need so many people?

**Comment by mprzygrodzki** [comment]
Hello everyone, i have a task to deploy BBB being scaleable in AWS as i can easily configure the needs in case of huge load on single machine, similiar just like the author of this issue - after reading all posts i see it's currently not supported but i would like to check this config behaviour from my end.I've got some problems running BBB as AutoScalingGroup. 

1) Firstly i've to mention that i'm building golden image with configured BBB and then creating snapshot for it. But, after launching new machines every node has different interal IP address that doesn't match the configuration ? How should i deal with this ? Is that ok to set 127.0.0.1 for the internal IP in configuration files ? While installing i'm running bbb-conf --setip bbb.mydomain.com --> this domain is pointed to external LoadBalancer which should route the traffic to the EC2.
Running bbb-conf --check displays internal IP from node on which BBB was built and it's different than on the currently running one. Is there any command which will directly update the config files fro freeswitch/sip/tomcat etc ?

2) Second question - when i'm going to launch let's say 3 nodes with the same configuration of BBB (each has it's own DB, same secrets etc) is this ok ? Or there should be separate node with MongoDB and all the instances connected to it ?
3) I would like to avoid deploying extra node with Scalelite, but if that's really needed i'll need to take a look with this kind of setup.

**Comment by congthang1** [comment]
Yes internal IP is my problem and I’m not yet have chance to configure all 127.0.0.1 as many place of configuration. For autoscaling on aws better use bbb docker on ECS cluster that will help you have share EFS volume or s3 and just need to give it ip of task. Even can use aws fargate to have external ip of each task. Then you need a sever api at Scalite, when a new task scale out or in it will tell Scalite cli to update its IP. Docker also can run on your ami too so can test this local first.

**Comment by congthang1** [comment]
One thing you need to consider is data transfer cost. In my case I see if 1 session enable both camera and micro may have about 4MB/minute/1. If you have big room with a lot people enable camera then the traffic cost can go over the server cost itself: eg. 100 people: (100*4*60/1024)*0.14 = $3.2/h. Consider to use aws Lightsail and not autoscaling but more big server with traffic almost free.

**Comment by mprzygrodzki** [comment]
@congthang1 At the time being will try building golden-image on EC2 (it's fully automated using Ansible - that's the main requirement for the project), and provide scripts that on each start of EC2 will update fields with local IP in all config files - will try this approach and see if it's fairly ok and then perform some tests running about 1-3 nodes in ASG behind LoadBalancer.

Simple question - for what kind of use NFS/S3 shared volume sie required ? To keep recordings ?

**Comment by congthang1** [comment]
Use EFS instead, its built-in on ec2 launch template. If you want to use s3 then need script to mount s3 with a library.

**Comment by opencubicles** [comment]
Hello everyone, we have been able to get an auto-scaling component (in AWS Lambda) along with scalelite  (from blindsidenetworks) load balancer up and running in our AWS Account. here is what we are following:

1.  Fully packaged AWS bigbluebutton AMI created from standard bbb-install.sh script
2. Any EC2 instance provisioned using above AWS AMI, will run a start-up script to change the domain configured for the bigbluebutton server while building AMI. Default Public DNS assigned by AWS was then fed to start-up script.  Reason for using default public DNS is because AWS has a limit of no. of elastic IPs you can get and mapping Elastic IPs to your domain etc in DNS server takes a bit of time to propagate. Using public DNS autogenerated by EC2 doesn't have any such issues!
3. We keep a fleet of about 20 such EC2 instances (AWS Limits, increasable) ready in AWS account but turned off. AWS doesn't bill you for turned off instance. (You get billed very little just for EBS volume)
4. We have a small lambda function written which continuously monitor our meeting scheduler MySQL DB (LMS moodle), looks for the number of servers already running in EC2, come up with the additional number of servers to be started/provisioned. Based on that Lambda fires another LAMDA which is solely responsible for turning on/off EC2 servers. 
5. a small listener which again sits in lambda monitor for any new server started and stopped and then accordingly run add/remove rake command them to scalelite LB [https://github.com/blindsidenetworks/scalelite]
6. We have a small EC2 always running which hosts the scalelite LB + another small server provisioned which is our monitoring server with prometheus and grafana installed. 
7. For storing recordings, We have configured AWS EFS which is then mounted on each bigbluebutton server via scalelite LB. AWS EFS is built to scale on-demand to petabytes and you pay per GB only when you actually store the recordings. 
8. Our only 24x7 cost incurred here is for always running scalelite LB and monitoring servers. Cost for servers hosting the meetings is always incurred when the meetings are happening. Auto-scaling component runs fully on serverless lambda function with cloud watch monitoring (for cron job).

**Comment by ffdixon** [comment]
@opencubicles Thanks for sharing this.  As you add more details are you able to share the source so others could do the same and contribute back?

**Comment by congthang1** [comment]
@opencubicles Does AWS public DNS come up with ssl?

**Comment by matiasilva** [comment]
In fact, it'd be nice to get some documentation going on this. Perhaps someone get started on drafting a pull request. My idea is to have this and many others as community-contributed guides.

**Comment by GhaziTriki** [comment]
I just found this and you might find it useful

https://github.com/toshke/big-blue-button-cloudformation-cfhl

**Comment by opencubicles** [comment]
> @opencubicles Does AWS public DNS come up with ssl?

Any Request to public DNS terminates on the bb server. So, SSL setup will be handled by lets-encrypt here.

**Comment by opencubicles** [comment]
> @opencubicles Thanks for sharing this. As you add more details are you able to share the source so others could do the same and contribute back?

@ffdixon  Planning to do that when we have some concrete setup in place. Is there any online BBB community meetups happening? It will be good to discuss initiatives like this there and see what community has to say?

**Comment by mprzygrodzki** [comment]
@opencubicles i'm also interested looking at your sources with the setup.

**Comment by opencubicles** [comment]
All, we are planning on to reduce the bandwidth cost as AWS data out cost is pretty high. Is there any way to extract Kurento workflow and host that part only on say digital ocean and keep everything else on AWS? this way video streams can be directly handled by Kurento hosted on DO which doesn't charge you on per GB basis.

**Comment by congthang1** [comment]
@opencubicles  on this repo they split all component of BBB https://github.com/alangecker/bigbluebutton-docker
You can check there they use this Kurento docker image: https://hub.docker.com/r/kurento/kurento-media-server/ on the docker-compose file.
So this docker image can be putting anywhere and just connect to it. 
But when doing this your autoscaling on AWS may not give value as most of the load is on Kurento!
Bandwidth cost on AWS really high, on aws you have an other option of Lightsail but its not good as DO. 
An other thing you can consider is a selfhost CDN like https://trafficcontrol.apache.org and put it on DO, but I'm not sure it can support webrtc.

**Comment by iSamof** [comment]
@congthang1  - thank you for sharing this valuable info.

**Comment by opencubicles** [comment]
@congthang1 Can you suggest some other low cost but reliable hosting partners like DO who has good amount bandwidth included in the instance cost ... but aren't charging for turned off instances/servers (similar to AWS)

**Comment by congthang1** [comment]
On AWS Lightsail you can have instance that include 4-7TB of traffic and it can have peer connection with your EC2 or can use AWS CLI.
But AWS Lightsail will charge turned off instance also, what you can do is instead of starting an existing instance, making a script to create instance from your snapshot and setup things after it created (use AWS CLI) or delete instance when scaling down. 
For AWS starting instance or creating new instance have quite similar timing, so you can benefit from traffic free with your AWS resources.

**Comment by opencubicles** [comment]
@congthang1 @mprzygrodzki  @matiasilva let me know if anyone of you want to catchup on a short call to brainstorm and come up with practical solution which can work for everyone. Autoscaling is key here which is missing piece in beautifully implement BBB stack!

**Comment by congthang1** [comment]
@opencubicles An idea is you can just keep your Autocaling on AWS, each instance is combination of a BBB EC2 instance + $5 Lighstail (2TB) with peer connection.
On lightsail you just need a nginx to proxy all traffic through EC2 bbb instance via internal IP (internal data transfer cost is free). This way DNS and SSL will be on Lightsail. You need to manage to have Scalelite connect to public DNS lightsail.
This scenario if working you just need to pay 5$/month for each lightsail to get 2TB traffic.

**Comment by vlasov01** [comment]
@congthang1 You should consider the following reported on [reddit ](https://www.reddit.com/r/webhosting/comments/8uty9y/has_anyone_used_amazon_lightsail_what_was_your/)"Be very wary about the undocumented CPU caps. If you use more than 40% CPU for a straight hour you'll burn your CPU credits and get throttled back very aggressively."

**Comment by congthang1** [comment]
> @congthang1 You should consider the following reported on [reddit ](https://www.reddit.com/r/webhosting/comments/8uty9y/has_anyone_used_amazon_lightsail_what_was_your/)"Be very wary about the undocumented CPU caps. If you use more than 40% CPU for a straight hour you'll burn your CPU credits and get throttled back very aggressively."

Yes I found that there is no way to stay on AWS with conference, the bandwidth will kill me. I decided to move to Digitalocean and will be making script on [Terraform](https://www.terraform.io).

**Comment by vlasov01** [comment]
> Digitalocean

@congthang1 Are they still charging powered off Droplets? Or you going to build your droplet every time you need to run it?

**Comment by congthang1** [comment]
I have a snapshot of bbb, our users need to schedule there class before 5 mins. It will build new droplet and script to add the registered IP to bbb (domain already pointed, ssl on snapshot with wildcard). And it will destroy the droplet when class end.

**Comment by iSamof** [comment]
@congthang1 and @vlasov01 - I see a lot of people are worried about the costs of stopped instances and therefore go through all kinds of work in order to create and delete instances.

We have used Prometeus.net for a very long time and are happy with them (low cost, very high reliability and great support) ... they have a cloud offering at iwStack.com ... which we have "not used" but they do not charge for stopped instances. I wonder if this will be of use for BBB.

Disclaimer: we have been a VPS customer of Prometeus for probably 8 years but have no other association with them.

**Comment by vlasov01** [comment]
> @congthang1 and @vlasov01 - I see a lot of people are worried about the costs of stopped instances and therefore go through all kinds of work in order to create and delete instances.
> 
> We have used Prometeus.net for a very long time and are happy with them (low cost, very high reliability and great support) ... they have a cloud offering at iwStack.com ... which we have "not used" but they do not charge for stopped instances. I wonder if this will be of use for BBB.
> 
> Disclaimer: we have been a VPS customer of Prometeus for probably 8 years but have no other association with them.

Thank you! It seems like a good option if you host your conferences for European audience. I'm not sure it will work for me as my audience is mostly in Canada.

**Comment by congthang1** [comment]
> > @congthang1 and @vlasov01 - I see a lot of people are worried about the costs of stopped instances and therefore go through all kinds of work in order to create and delete instances.
> > We have used Prometeus.net for a very long time and are happy with them (low cost, very high reliability and great support) ... they have a cloud offering at iwStack.com ... which we have "not used" but they do not charge for stopped instances. I wonder if this will be of use for BBB.
> > Disclaimer: we have been a VPS customer of Prometeus for probably 8 years but have no other association with them.
> 
> Thank you! It seems like a good option if you host your conferences for European audience. I'm not sure it will work for me as my audience is mostly in Canada.

Found this interesting Repo they have autoscale BBB on IONOS provider: https://github.com/hpi-schul-cloud/bbb-deployment
I see they have cheep VPS with unlimited bandwidth (400Mbps) and dedicated server (start at 4 core 16GB for $70/month hourly and 1Gbps network up) located in US and EU so in Canada you can use US. 
I have not yet tried to deploy this repo as I'm in AS and its only for IONOS (need to convert to other provider using Teraform)

**Comment by vlasov01** [comment]
> > > @congthang1 and @vlasov01 - I see a lot of people are worried about the costs of stopped instances and therefore go through all kinds of work in order to create and delete instances.
> > > We have used Prometeus.net for a very long time and are happy with them (low cost, very high reliability and great support) ... they have a cloud offering at iwStack.com ... which we have "not used" but they do not charge for stopped instances. I wonder if this will be of use for BBB.
> > > Disclaimer: we have been a VPS customer of Prometeus for probably 8 years but have no other association with them.
> > 
> > 
> > Thank you! It seems like a good option if you host your conferences for European audience. I'm not sure it will work for me as my audience is mostly in Canada.
> 
> Found this interesting Repo they have autoscale BBB on IONOS provider: https://github.com/hpi-schul-cloud/bbb-deployment
> I see they have cheep VPS with unlimited bandwidth (400Mbps) and dedicated server (start at 4 core 16GB for $70/month hourly and 1Gbps network up) located in US and EU so in Canada you can use US.
> I have not yet tried to deploy this repo as I'm in Asia and its only for IONOS and IONOS doesn't have server in Asia (need to convert to other provider using Teraform)

Thank you! It is an interesting project. But I need to understand how it handles recordings as the architecture diagram is a high-level and does not explaining how they managing storage layer.
IONOS is relying on their partner Anexio for their NJ data center, which is quite close for my use case. But I can't find SLA for their cloud servers, availability zones. Their service availability report is quite short and is not very detailed. As well I'm not sure if you can scale down with OPNAS as per an [open issue](https://github.com/hpi-schul-cloud/bbb-deployment/issues/3) on the project:  .

**Comment by congthang1** [comment]
Found an idea that can make BBB autoscale on kubernetes (K8S) without Scalelite:

- An ingress controller with "Least Connection" routing and sticky to the node.
- Bigbluebutton is on a Statefulset and each with a service name on k8s.
- A nginx front to route traffic to bbb service name.
---
If there are 2 bbb instance: bbb-1 and bbb-2 running.
A new api creating room will come to the less load BBB statefulset by "Least Connection", eg: bbb-1 selected .
The api response should include bbb service name as params: node=bbb1.
The client join api will bring back this params: eg: &node=bbb-1 then nginx will route this url with &node=bbb-1 to bbb-1 service.
The client url with token also need to include this params too: &node=bbb-1 to make sure user will go same instance.

When first time this user connect to this instance, the Sticky session of k8s will stick any request from this cookie to that node. So user stay on that bbb-1 without any modify needed.

The bbb instance can be autoscale using Kubernetes Horizontal Pod Autoscale (HPA). 

Is this possible?

**Comment by congthang1** [comment]
Successfully scale bigbluebutton on K8S. Check my repo https://github.com/congthang1/bigbluebutton-kubernetes-autoscale
